### ISSUE TITLE
Data Distillery: Change format of CodeID property of Code node

### ISSUE THREAD

**Issue by AlanSimmons** [dbmi-pitt/ubkg#20]
# Issue
The Code node currently has the following properties:

- SAB
- CODE
- CodeID

The **CodeID** property is currently in format _SAB_ _space_  _code in SAB_ This format is different from the way that entities are represented in the EMBL-EBI service, which is _SAB_:_code in SAB_.

![image](https://user-images.githubusercontent.com/10928372/206593104-7edaeeaf-bb2a-4dc8-8852-715971b78df9.png)

<img width="449" alt="image" src="https://user-images.githubusercontent.com/10928372/206593150-9fda46ef-48e3-4634-afea-080d7e4193b3.png">

# Solution

Change how the generation framework represents **CodeId** in the Code node.

**Comment by AlanSimmons** [comment]
# Analysis
Changing the format of the CodeID property would require fundamental changes to a number of elements of both the Source Framework (that builds the original UMLS CSVs) and the Generation Framework (that appends to the UMLS CSVs to form the ontology CSVs). 

Following is a discussion of things to consider.

1. Current state
2. Potential branching issue with API

# Current state
By default, the two frameworks of the UBKG (Source Framework and Generation Framework) assume that codes from standard biomedical ontologies are stored separately from the SABs (Source Abbreviations, or ontology namespaces) to which they belong. This results in Code nodes with CodeId property in the format SAB Code--i.e., delimited with a space (e.g., "SNOMEDCT_US 156072005").

## Code representation in UMLS CSVs
**CODES.CSV**

CodeID:ID |SAB |CODE
--- | --- | ---
MSH D015056| MSH| D015056

**CODE-SUIs.CSV**
:END_ID |:START_ID |:TYPE|CUI
--- | --- | --- | ---
S0033295|MSH D015060|ET|C0000039

**CUI-CODES.CSV**
:START_ID|:END_ID
--- | ---
C0000039| LNC LP15542-1

## neo4j-admin import
The neo4j-admin import process runs on the ontology CSVs with command:

```RUN ./neo4j-admin import --verbose --database=ontology --nodes=Semantic="${IMPORT}/TUIs.csv" --nodes=Concept="${IMPORT}/CUIs.csv" --nodes=Code="${IMPORT}/CODEs.csv" --nodes=Term="${IMPORT}/SUIs.csv" --nodes=Definition="${IMPORT}/DEFs.csv" --relationships=ISA_STY="${IMPORT}/TUIrel.csv" --relationships=STY="${IMPORT}/CUI-TUIs.csv" --relationships="${IMPORT}/CUI-CUIs.csv" --relationships=CODE="${IMPORT}/CUI-CODEs.csv" --relationships="${IMPORT}/CODE-SUIs.csv" --relationships=PREF_TERM="${IMPORT}/CUI-SUIs.csv" --relationships=DEF="${IMPORT}/DEFrel.csv" --skip-bad-relationships --skip-duplicate-nodes```

The parameters relative to codes are:
* ```--nodes=Code="${IMPORT}/CODEs.csv"```
* ```--relationships=CODE="${IMPORT}/CUI-CODEs.csv" ```
* ```--relationships="${IMPORT}/CODE-SUIs.csv"```

##  Source framework script
The script that exports UMLS data from the DBMI data mart formats CodeIds as shown above.

## OWLNETS-UMLS-GRAPH conversion script

1. The script employs merges against the *CUI-CODEs.CSV* file to associate codes with CUIs--in particular, the **:END_ID** column.
3. The script appends to the 3 CSV files (CODES.CSV, CODE-SUIs.CSV, CUI-CODEs.CSV).

##  Ontology API
Endpoints to the API that queries ontology Knowledge Graph have responses that include the codes. The **codes** endpoint accepts a code ID with a space as a parameter--e.g.,
https://ontology.api.hubmapconsortium.org/codes/SNOMEDCT_US%20254837009/concepts

# Potential branching issue with API
The HubMAP implementation of an ontology KG was based on the CodeID not including the SAB by default. 

It would be possible to convert the CodeID as a separate step between generation and import into neo4j to support the Data Distillery. However, this could cause a branching issue with respect to the API. At the least, changes to the CodeID would entail testing of the API.

**Comment by AlanSimmons** [comment]
@computationdoc Following are the obvious places where framework scripts would need to change. 

## Graph extract script

[Source]( https://github.com/dbmi-pitt/ubkg/blob/main/source_framework/UMLS-Graph-Extracts.py)

Lines: 

- 106 (queries for export to CODEs.CSV)
- 119 (queries for export to CUI-CODEs.CSV)
- 143 (queries for export to CODE-SUIs.CSV)

## OWLNETS-UMLS-GRAPH-12.py

[Source](https://github.com/dbmi-pitt/ubkg/blob/main/scripts/Jonathan/OWLNETS-UMLS-GRAPH-12.py)

Lines:

- 610 (merge node_metadata with loaded CUI-CODEs.CSV)
- 803-811 (appends to CODEs.CSV)
- 845-862 (appends to CUI-CODEs.csv)
- possibly in the code that appends to CODE-SUIs.CSV

## API
It does not appear at first glance that the code behind the code endpoint would have to change.

[Source](https://github.com/dbmi-pitt/ubkg/blob/main/api/openapi_server/managers/neo4j_manager.py)
Lines:

- 113 - 130 (codes_code_id_codes_get)
- 132 - 149 (codes_code_id_concepts_get)

**Comment by computationdoc** [comment]
There are a number of considerations (here are a few top of head):

1. HGNC, GO, others submitting to UMLS may be doing it that way to avoid concerns of leading zeros - if we break it out we recreate the problem they’ve solved in spreadsheets

2. If we force everyone to use the same approach we take away flexibility the ontology owners currently enjoy

3. The space was selected as a “reserved character” so no matter what folks do in future we’re covered (except if folks start using spaces in SABs or CODEs - seems rather unlikely)

4. We can allow whatever folks want in ingest node and edge files because that is a distinct ingest software - however, we probably should not induce restrictions unless absolutely necessary that diverge from OWLNETS’ approach - so that those OWLNETs outputs and our custom both work easily…

5. APIs can handle anything “out” of the database

6. We “may” want to consider API-approach to ingest into database - what would be the implications? (I’m thinking of a robust editor at some future - not just simple knowledge editor)

7. Not sure of all the implications but what about use of colon in URLs (probably we URL encode anyway)

8. Not sure of all the implications but what about use of underscore in current general approach to ontology-lookups on various sources’ URIs

Bottom line: tread carefully in new territory

Cheers,

JCS

**Comment by AlanSimmons** [comment]
# Final Answer

## Short answer
We cannot change the format of the CodeID; however, we do offer ways to work with the current format in both Cypher queries and in API calls.

## Longer answer
@computationdoc and I discussed the issue at length.

### Why we need to keep the space as a delimiter
There are a number of reasons why we believe that we should use a space as the delimiter between SAB and code in the CodeId property for Code nodes in the UBKG, including:

1. There is not a standard way to represent codes in all of the ontologies with which we have worked. Some use colons as delimiters; some use underscores. We chose to use the space as a "reserved character" for delimiting.
2. Some ontologies (including HPO, GO, and HGNC) concatenate SAB and code with colons in the UMLS to prevent truncation of leading zeros in codes. Some SAB strings include underscores, and might include colons. All of this means that reformatting the original codes might cause issues later--e.g., we might create CodeIds with two colons.
4. We believe that formatting CodeId with a space delimiter better conforms to codes ingested from ontologies consolidated by the UMLs.

In other words, we assert that the space delimiter is a necessary part of the database design. We actually plan to describe our rationale in a later paper.

 ### How to handle this in Cypher queries
The question then would seem to be: **How can I query the UBKG for codes using a parameter in format _SAB:Code_?**

We recommend using WITH clauses that use string functions as prefixes to Cypher queries.

For example, to search the existing UBKG for the code corresponding to UBERON:004221, include as prefix

```
WITH split("UBERON:0004221", ":") AS x  
MATCH (myCode:Code{SAB:x[0], CODE:x[1]}) 
WITH myCode 
RETURN myCode.CodeID 
```

Example return:

![image](https://user-images.githubusercontent.com/10928372/210415952-c0f02e80-21a4-481b-9e48-f3aea37ce276.png)

### How to handle this in API calls
At the moment, the API currently passes the parameter to the code endpoint without parsing--i.e., it assumes that the parameter conforms to the format of SAB CODE. It would be possible to modify the API so that it could accept code strings that included colon delimiters. This would not require changes to the underlying knowledge graph database.
